toolkit/tools/pkg/imagecustomizerlib/partitionutils.go

```diff
@@ -352,24 +392,164 @@ func findPartition(mountIdType imagecustomizerapi.MountIdentifierType, mountId s
+		matches, err := checkExtendedDevPartition(partition, partitions, buildDir)
+		if err != nil {
+			return "", err
+		}
+		if matches {
+			return partition.PartUuid, nil
+		}
+	}
+	return "", fmt.Errorf("unable to find PARTUUID for /dev path")
+}
+
+func checkExtendedDevPartition(partition diskutils.PartitionInfo, partitions []diskutils.PartitionInfo,
+	buildDir string,
+) (bool, error) {
+	cmdline, err := extractKernelCmdline(partitions, buildDir)
+	if err != nil {
+		return false, err
+	}
+
+	verityPartUUID, err := extractVerityRootPartUUID(cmdline)
+	if err != nil {
+		return false, err
+	}
+
+	return partition.PartUuid == verityPartUUID, nil
+}
+
+func extractKernelCmdline(partitions []diskutils.PartitionInfo, buildDir string) (string, error) {
+	espPartition, err := findSystemBootPartition(partitions)
+	if err != nil {
+		return "", fmt.Errorf("failed to find ESP partition: %w", err)
+	}
+
+	bootPartition, err := findBootPartitionFromEsp(espPartition, partitions, buildDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to find boot partition: %w", err)
+	}
+
+	tmpDirEsp := filepath.Join(buildDir, tmpEspPartitionDirName)
+	espPartitionMount, err := safemount.NewMount(espPartition.Path, tmpDirEsp, espPartition.FileSystemType, unix.MS_RDONLY, "", true)
+	if err != nil {
+		return "", fmt.Errorf("failed to mount ESP partition (%s):\n%w", espPartition.Path, err)
+	}
+	defer espPartitionMount.Close()
+
+	espLinuxPath := filepath.Join(tmpDirEsp, UkiOutputDir)
+	ukiFiles, err := filepath.Glob(filepath.Join(espLinuxPath, "vmlinuz-*.efi"))
+	if err != nil {
+		return "", fmt.Errorf("failed to search for UKI images in ESP partition:\n%w", err)
+	}
+
+	if len(ukiFiles) > 0 {
+		cmdlinePath := filepath.Join(buildDir, "cmdline.txt")
+		_, _, err := shell.Execute("objcopy", "--dump-section", ".cmdline="+cmdlinePath, ukiFiles[0])
+		if err != nil {
+			return "", fmt.Errorf("failed to dump kernel cmdline args from UKI:\n%w", err)
+		}
+
+		cmdlineContent, err := os.ReadFile(cmdlinePath)
+		if err != nil {
+			return "", fmt.Errorf("failed to read kernel cmdline args from dumped file:\n%w", err)
+		}
+
+		err = espPartitionMount.CleanClose()
+		if err != nil {
+			return "", fmt.Errorf("failed to close bootPartitionMount:\n%w", err)
+		}
+
+		return string(cmdlineContent), nil
+	}
+
+	tmpDirBoot := filepath.Join(buildDir, tmpBootPartitionDirName)
+	bootPartitionMount, err := safemount.NewMount(bootPartition.Path, tmpDirBoot, bootPartition.FileSystemType, unix.MS_RDONLY, "", true)
+	if err != nil {
+		return "", fmt.Errorf("failed to mount boot partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer bootPartitionMount.Close()
+
+	grubCfgPath := filepath.Join(tmpDirBoot, "/grub2/grub.cfg")
+	kernelToArgs, err := extractKernelToArgsFromGrub(grubCfgPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to extract kernel arguments from grub.cfg:\n%w", err)
+	}
+
+	var combinedArgs []string
+	for _, args := range kernelToArgs {
```

Comment:
Shouldn't you only use the args from a single kernel version, instead of all of them?